pkg/daemon/ceph/multus/utils.go

```diff
@@ -0,0 +1,358 @@
+			id, err := strconv.Atoi(idStrs[1])
+			if err != nil {
+				log.Fatal(err)
+			}
+			if id > linkNumber {
+				linkNumber = id
+			}
+		}
+	}
+	linkNumber += 1
+
+	newLinkName = fmt.Sprintf("%s%d", ifBase, linkNumber)
+
+	return newLinkName, nil
+}
+
+func migrateInterface(hostNS, holderNS ns.NetNS, ogLinkName, newLinkName string) error {
+	return holderNS.Do(func(ns ns.NetNS) error {
+		link, err := netlink.LinkByName(ogLinkName)
+		if err != nil {
+			return err
+		}
+
+		logger.Info("setting multus link down to be renamed")
+		if err := netlink.LinkSetDown(link); err != nil {
+			return err
+		}
+
+		logger.Infof("renaming multus link to %s", newLinkName)
+		if err := netlink.LinkSetName(link, newLinkName); err != nil {
+			return err
+		}
+
+		// After renaming the link, the link object must be updated or netlink will get confused.
+		link, err = netlink.LinkByName(newLinkName)
+		if err != nil {
+			return err
+		}
+
+		logger.Info("moving the multus interface to the host network namespace")
+		if err = netlink.LinkSetNsFd(link, int(hostNS.Fd())); err != nil {
+			return err
+		}
+		return nil
+	})
+}
+
+func setupInterface(mLinkName string, multusIP netlink.Addr) error {
+	link, err := netlink.LinkByName(mLinkName)
+	if err != nil {
+		return err
+	}
+
+	// The IP address label must be changed to the new interface name
+	// for the AddrAdd call to succeed.
+	multusIP.Label = mLinkName
+
+	if err := netlink.AddrAdd(link, &multusIP); err != nil {
+		return err
+	}
+
+	logger.Info("setting link up")
+	if err := netlink.LinkSetUp(link); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkMigration(multusIpStr string) (bool, string, error) {
+	var migrated bool
+	var linkName string
+
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		return migrated, linkName, err
+	}
+
+	for _, iface := range interfaces {
+		link, err := netlink.LinkByName(iface.Name)
+		if err != nil {
+			return migrated, linkName, err
+		}
+		if link == nil {
+			return migrated, linkName, errors.New("link not found")
+		}
+
+		addrs, err := netlink.AddrList(link, 0)
+		if err != nil {
+			return migrated, linkName, err
+		}
+
+		for _, addr := range addrs {
+			if addr.IP.String() == multusIpStr {
+				migrated = true
+				linkAttrs := link.Attrs()
+				if linkAttrs != nil {
+					linkName = linkAttrs.Name
+				}
+				return migrated, linkName, nil
+			}
+		}
+	}
+
+	return migrated, linkName, nil
+}
+
+func determineMultusIPConfig(holderNS ns.NetNS, multusIP, multusLinkName string) (netlink.Addr, error) {
+	var mAddr netlink.Addr
+	var addrFound bool
+
+	err := holderNS.Do(func(ns ns.NetNS) error {
+		logger.Info("finding the multus network connected link")
+		link, err := netlink.LinkByName(multusLinkName)
+		if err != nil {
+			return err
+		}
+
+		logger.Info("determining the IP address of the multus link")
+		addrs, err := netlink.AddrList(link, 0)
+		if err != nil {
+			log.Fatal(err)
```

Comment:
Why `Fatal` here? That's not how we have patterned our error handling in Rook. We want to be able to gracefully recover from failures and retry if it makes sense.